wakuv2/waku.go

```diff
@@ -1015,14 +1079,85 @@ func (w *Waku) Send(pubsubTopic string, msg *pb.WakuMessage) ([]byte, error) {
 	alreadyCached := w.envelopeCache.Has(gethcommon.BytesToHash(envelope.Hash().Bytes()))
 	w.poolMu.Unlock()
 	if !alreadyCached {
-		recvMessage := common.NewReceivedMessage(envelope, common.RelayedMessageType)
+		recvMessage := common.NewReceivedMessage(envelope, common.SendMessageType)
 		w.postEvent(recvMessage) // notify the local node about the new message
 		w.addEnvelope(recvMessage)
 	}
 
 	return envelope.Hash().Bytes(), nil
 }
 
+// ctx, peer, r.PubsubTopic, contentTopics, uint64(r.From), uint64(r.To), options, processEnvelopes
+func (w *Waku) messageHashBasedQuery(ctx context.Context, hashes []gethcommon.Hash, pubsubTopic string) []gethcommon.Hash {
+	selectedPeer := w.storePeerID
+	if selectedPeer == "" {
+		selectedPeers, err := w.node.PeerManager().SelectPeers(
+			peermanager.PeerSelectionCriteria{
+				SelectionType: peermanager.Automatic,
+				Proto:         store.StoreQueryID_v300,
+				PubsubTopics:  []string{pubsubTopic},
+				Ctx:           ctx,
+			},
+		)
+		if err != nil {
+			w.logger.Error("could not select peers", zap.Error(err))
+			return []gethcommon.Hash{}
+		}
+		selectedPeer = selectedPeers[0]
+	}
+
+	var opts []store.RequestOption
+	requestID := protocol.GenerateRequestID()
+	opts = append(opts, store.WithRequestID(requestID))
+	opts = append(opts, store.WithPeer(selectedPeer))
+
+	messageHashes := make([]pb.MessageHash, len(hashes))
+	for i, hash := range hashes {
+		messageHashes[i] = pb.ToMessageHash(hash.Bytes())
+	}
+
+	w.logger.Debug("store.queryByHash request", zap.String("requestID", hexutil.Encode(requestID)), zap.String("peerID", selectedPeer.String()), zap.Any("messageHashes", messageHashes))
+
+	result, err := w.node.Store().QueryByHash(ctx, messageHashes, opts...)
```

Comment:
Since we're not using the content of the message, it's more efficient to add to `opts` the option `store.IncludeData(false)`. There's a refactor over storev3 being worked on that will make this type of queries more efficient (See https://github.com/waku-org/nwaku/issues/2637), also, not including the data means less bandwidth used